src/main/java/org/xbill/DNS/DNSSession.java

```diff
@@ -0,0 +1,129 @@
+package org.xbill.DNS;
+
+import static java.lang.String.format;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import org.xbill.DNS.exceptions.AdditionalDetail;
+import org.xbill.DNS.exceptions.InvalidZoneDataException;
+import org.xbill.DNS.exceptions.LookupFailedException;
+import org.xbill.DNS.exceptions.RedirectOverflowException;
+import org.xbill.DNS.exceptions.ServerFailedException;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+public class DNSSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+
+  private final Resolver resolver;
+  private volatile int maxRedirects = DEFAULT_MAX_ITERATIONS;
+
+  /**
+   * Construct a DNSSession using the provided Resolver to lookup records.
+   *
+   * @param resolver the Resolver instance to backing this DNSSession.
+   */
+  DNSSession(Resolver resolver) {
+    this.resolver = resolver;
+  }
+
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * RedirectOverflowException.
+   *
+   * @param maxRedirects the maximum number of allowed redirections.
+   */
+  public void setMaxRedirects(int maxRedirects) {
+    this.maxRedirects = maxRedirects;
+  }
+
+  /**
+   * Make an asynchronous lookup of the provided name.
+   *
+   * @param name the name to look up.
+   * @param type the type to look up, values should correspond to constants in Type.
```

Comment:
Link to classes with `{@link Type}` (in other places too)

Reply:
Adding..